src/test/webServer.ts

```diff
@@ -0,0 +1,140 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/* eslint-disable no-console */
+import { ChildProcess, exec } from 'child_process';
+import net from 'net';
+import os from 'os';
+import stream from 'stream';
+import kill from 'tree-kill';
+import { monotonicTime, raceAgainstDeadline } from './util';
+import { WebServerConfig } from '../../types/test';
+import { assert } from '../utils/utils';
+
+const DEFAULT_ENVIRONMENT_VARIABLES = {
+  'BROWSER': 'none', // Disable that create-react-app will open the page in the browser
+};
+
+const newProcessLogPrefixer = () => new stream.Transform({
+  transform(this: stream.Transform, chunk: any, encoding: string, callback: stream.TransformCallback) {
+    this.push(chunk.toString().split(os.EOL).map((line: string): string => line ? `[Playwright WebServer] ${line}` : line).join(os.EOL));
+    callback();
+  },
+});
+
+export class WebServer {
+  private _process?: ChildProcess;
+  private _port?: number;
+  private _processExitedWithNonZeroStatusCode!: Promise<any>;
+  private _config: Omit<WebServerConfig, 'port'>;
+  constructor(config: WebServerConfig) {
+    this._config = config;
+    this._port = config.port;
+  }
+
+  public static async create(config: WebServerConfig): Promise<WebServer> {
+    const webServer = new WebServer(config);
+    if (webServer._port)
+      await webServer._verifyFreePort();
+    try {
+      await webServer._startWebServer();
+      await webServer._waitForAvailability();
+    } catch (error) {
+      await webServer.kill();
+      throw error;
+    }
+    process.env.PLAYWRIGHT_WEB_SERVER_URL = `http://localhost:${webServer._port}`;
+    return webServer;
+  }
+
+  private async _verifyFreePort() {
+    assert(this._port);
+    const cancellationToken = { canceled: false };
+    const portIsUsed = await Promise.race([
+      new Promise(resolve => setTimeout(() => resolve(false), 100)),
+      waitForSocket(this._port, 100, cancellationToken),
+    ]);
+    cancellationToken.canceled = true;
+    if (portIsUsed)
+      throw new Error(`Port ${this._port} is used, make sure that nothing is running on the port`);
```

Comment:
Is this just a nicety? I think starting the server will just fail if we do not verify the port. It is also racy...

Reply:
It was for the case when the user specifies a command and manually a port. We would need to ensure then that the process is running while the port is accepting connections which was otherwise more racy than that in my mind.